anospp_analysis/nn.py

```diff
@@ -0,0 +1,501 @@
+            abs_kmer_dist = np.abs(kmers[cand.target,cand_parsed_seqid[1],:] - kmers[cand.target,possible_source,:]).sum()
+            if abs_kmer_dist<threshold:
+                error_seqs.append(cand.seqid)
+                break
+    
+    logging.info(f'identified {len(error_seqs)} error sequences')
+
+    return(error_seqs)
+
+def recompute_haplotype_coverage(hap_df):
+    hap_df = hap_df.drop(["total_reads", "reads_fraction", "nalleles"], axis=1)
+
+    hap_df['total_reads'] = hap_df.groupby(by=['sample_id', 'target']) \
+            ['reads'].transform('sum')
+
+    hap_df['reads_fraction'] = hap_df['reads'] / hap_df['total_reads']
+
+    hap_df['nalleles'] = hap_df.groupby(by=['sample_id', 'target']) \
+            ['consensus'].transform('nunique')
+    return(hap_df)
+
+def compute_kmer_distance(kmers, ref_kmers, tgt, qidx, refidx):
+    '''
+    compute k-mer distance between query kmer count
+    and ref kmer count(s)
+    returns absolute and normalised distance
+    '''
+    #identify k-mer mismatches
+    diff = np.abs(ref_kmers[tgt, refidx, :] - kmers[tgt, qidx, :])
+    total = np.sum(ref_kmers[tgt, refidx, :] + kmers[tgt, qidx, :], axis=1)
+    #sum to get absolute distance
+    dist = np.sum(diff, axis=1)
+    #normalise
+    norm_dist = dist / total
+
+    return(dist, norm_dist)
+
+def find_nn_unique_haps(non_error_hap_df, kmers, ref_hap_df, ref_kmers):
+    '''
+    identify the nearest neighbours of the unique haplotypes in the reference dataset 
+    '''
+    #get idxs occupied for each target
+    parsed_ref_seqids = parse_seqid(ref_hap_df.seqid.drop_duplicates())
+    ref_idxs_per_target = parsed_ref_seqids.groupby(0)[1].unique()
+
+    nndict = dict()
+
+    logging.info(f"identifying nearest neighbours for {non_error_hap_df.seqid.nunique()} unique haplotypes")
+    
+    #loop through unique haplotypes
+    unique_seqids = non_error_hap_df.seqid.unique()
+    for seqid in unique_seqids:
+        tgt, qidx = parse_seqid(seqid).loc[0,0], parse_seqid(seqid).loc[0,1]
+        #compute distance between focal hap and all same target haps in ref index
+        dist, norm_dist = compute_kmer_distance(kmers, ref_kmers, tgt, qidx, ref_idxs_per_target[tgt])
+        #Find nearest neighbours
+        nn_qidx = ref_idxs_per_target[tgt][norm_dist==norm_dist.min()]
+        #include in dict
+        nndict[seqid] = (nn_qidx, norm_dist.min())
+
+    return(nndict)
+
+def perform_nn_assignment_samples(non_error_hap_df, ref_hap_df, nndict, af_c, af_i, af_f,\
+                                  normalisation):
+    '''
+    The main NN assignment function
+    it outputs three dataframes containing the assignment proportions to each species-group for the three levels
+    '''
+    #get samples with at least 10 targets
+    test_samples = non_error_hap_df.groupby('sample_id').filter(lambda x: x['target'].nunique() >=10)['sample_id'].unique()
+
+    logging.info(f'performing NN assignment for {len(test_samples)} samples')
```

Comment:
I see your point, but then I'd llike to report the total number of samples in that run, because `non_error_hap_df['sample_id'].nunique()` are only the samples with some mosquito targets -- i.e. the samples with no mosquito targets passing our thresholds are not present in `non_error_hap_df`. If you think it's worth it for just reporting purposes I can pass `comb_stats_df` or its shape to this function, but to me it seems cleaner not to. 
I have added an info line to the nn function stating the total number of samples in the run.